src/data/sum/order.lean

```diff
@@ -0,0 +1,491 @@
+lift_rel_inl_inl
+
+@[simp] lemma inr_lt_inr_iff [has_lt α] [has_lt β] {a b : β} : (inr a : α ⊕ β) < inr b ↔ a < b :=
+lift_rel_inr_inr
+
+@[simp] lemma not_inl_le_inr [has_le α] [has_le β] {a : α} {b : β} : ¬ inl b ≤ inr a :=
+not_lift_rel_inl_inr
+
+@[simp] lemma not_inl_lt_inr [has_lt α] [has_lt β] {a : α} {b : β} : ¬ inl b < inr a :=
+not_lift_rel_inl_inr
+
+@[simp] lemma not_inr_le_inl [has_le α] [has_le β] {a : α} {b : β} : ¬ inr b ≤ inl a :=
+not_lift_rel_inr_inl
+
+@[simp] lemma not_inr_lt_inl [has_lt α] [has_lt β] {a : α} {b : β} : ¬ inr b < inl a :=
+not_lift_rel_inr_inl
+
+section preorder
+variables [preorder α] [preorder β]
+
+instance : preorder (α ⊕ β) :=
+{ le_refl := λ _, refl _,
+  le_trans := λ _ _ _, trans,
+  lt_iff_le_not_le := λ a b, begin
+    refine ⟨λ hab, ⟨hab.mono (λ _ _, le_of_lt) (λ _ _, le_of_lt), _⟩, _⟩,
+    { rintro (⟨b, a, hba⟩ | ⟨b, a, hba⟩),
+      { exact hba.not_lt (inl_lt_inl_iff.1 hab) },
+      { exact hba.not_lt (inr_lt_inr_iff.1 hab) } },
+    { rintro ⟨⟨a, b, hab⟩ | ⟨a, b, hab⟩, hba⟩,
+      { exact lift_rel.inl (hab.lt_of_not_le $ λ h, hba $ lift_rel.inl h) },
+      { exact lift_rel.inr (hab.lt_of_not_le $ λ h, hba $ lift_rel.inr h) } }
+  end,
+  .. sum.has_le, .. sum.has_lt }
+
+lemma inl_mono : monotone (inl : α → α ⊕ β) := λ a b, lift_rel.inl
+lemma inr_mono : monotone (inr : β → α ⊕ β) := λ a b, lift_rel.inr
+lemma inl_strict_mono : strict_mono (inl : α → α ⊕ β) := λ a b, lift_rel.inl
+lemma inr_strict_mono : strict_mono (inr : β → α ⊕ β) := λ a b, lift_rel.inr
+
+end preorder
+
+instance [partial_order α] [partial_order β] : partial_order (α ⊕ β) :=
+{ le_antisymm := λ _ _, antisymm,
+  .. sum.preorder }
+
+instance no_bot_order [has_lt α] [has_lt β] [no_bot_order α] [no_bot_order β] :
+  no_bot_order (α ⊕ β) :=
+⟨λ a, match a with
+| inl a := let ⟨b, h⟩ := no_bot a in ⟨inl b, inl_lt_inl_iff.2 h⟩
+| inr a := let ⟨b, h⟩ := no_bot a in ⟨inr b, inr_lt_inr_iff.2 h⟩
+end⟩
+
+instance no_top_order [has_lt α] [has_lt β] [no_top_order α] [no_top_order β] :
+  no_top_order (α ⊕ β) :=
+⟨λ a, match a with
+| inl a := let ⟨b, h⟩ := no_top a in ⟨inl b, inl_lt_inl_iff.2 h⟩
+| inr a := let ⟨b, h⟩ := no_top a in ⟨inr b, inr_lt_inr_iff.2 h⟩
+end⟩
+
+@[simp] lemma no_bot_order_iff [has_lt α] [has_lt β] :
+  no_bot_order (α ⊕ β) ↔ no_bot_order α ∧ no_bot_order β :=
+⟨by { introI, exact ⟨⟨λ a, begin
+  obtain ⟨b, h⟩ := no_bot (inl a : α ⊕ β),
+  cases b,
+  { exact ⟨b, inl_lt_inl_iff.1 h⟩ },
+  { exact (not_inr_lt_inl h).elim }
+end⟩, ⟨λ a, begin
+  obtain ⟨b, h⟩ := no_bot (inr a : α ⊕ β),
+  cases b,
+  { exact (not_inl_lt_inr h).elim },
+  { exact ⟨b, inr_lt_inr_iff.1 h⟩ }
+end⟩⟩ }, λ h, @sum.no_bot_order _ _ _ _ h.1 h.2⟩
+
+@[simp] lemma no_top_order_iff [has_lt α] [has_lt β] :
+  no_top_order (α ⊕ β) ↔ no_top_order α ∧ no_top_order β :=
+⟨by { introI, exact ⟨⟨λ a, begin
+  obtain ⟨b, h⟩ := no_top (inl a : α ⊕ β),
+  cases b,
+  { exact ⟨b, inl_lt_inl_iff.1 h⟩ },
+  { exact (not_inl_lt_inr h).elim }
```

Comment:
golfed:
```suggestion
  obtain ⟨bl | br, h⟩ := no_top (inl a : α ⊕ β),
  { exact ⟨bl, inl_lt_inl_iff.1 h⟩ },
  { exact (not_inl_lt_inr h).elim }
```
this can probably be applied in a few more places too.